plugins/optimization-detective/debug.php

```diff
@@ -0,0 +1,226 @@
+}
+
+add_action( 'od_register_tag_visitors', 'od_debug_register_tag_visitors', PHP_INT_MAX );
+
+/**
+ * Filters additional properties for the element item schema for Optimization Detective.
+ *
+ * @since n.e.x.t
+ *
+ * @param array<string, array{type: string}> $additional_properties Additional properties.
+ * @return array<string, array{type: string}> Additional properties.
+ */
+function od_debug_add_inp_schema_properties( array $additional_properties ): array {
+	$additional_properties['inpData'] = array(
+		'description' => __( 'INP metrics', 'optimization-detective' ),
+		'type'        => 'array',
+
+		/*
+		 * All extended properties must be optional so that URL Metrics are not all immediately invalidated once an extension is deactivated.
+		 * Also, no INP data will be sent if the user never interacted with the page.
+		 */
+		'required'    => false,
+		'items'       => array(
+			'type'                 => 'object',
+			'required'             => true,
+			'properties'           => array(
+				'value'             => array(
+					'type'     => 'number',
+					'required' => true,
+				),
+				'rating'            => array(
+					'type'     => 'string',
+					'enum'     => array( 'good', 'needs-improvement', 'poor' ),
+					'required' => true,
+				),
+				'interactionTarget' => array(
+					'type'     => 'string',
+					'required' => true,
+				),
+			),
+			'additionalProperties' => true,
+		),
+	);
+	return $additional_properties;
+}
+
+add_filter( 'od_url_metric_schema_root_additional_properties', 'od_debug_add_inp_schema_properties' );
+
+/**
+ * Adds a new admin bar menu item for Optimization Detective debug mode.
+ *
+ * @since n.e.x.t
+ *
+ * @param WP_Admin_Bar $wp_admin_bar The WP_Admin_Bar instance, passed by reference.
+ */
+function od_debug_add_admin_bar_menu_item( WP_Admin_Bar &$wp_admin_bar ): void {
+	if ( ! current_user_can( 'customize' ) && ! wp_is_development_mode( 'plugin' ) ) {
+		return;
+	}
+
+	if ( ! is_admin_bar_showing() ) {
+		return;
+	}
+
+	if ( is_admin() ) {
+		return;
+	}
+
+	$wp_admin_bar->add_menu(
+		array(
+			'id'     => 'optimization-detective-debug',
+			'parent' => null,
+			'group'  => null,
+			'title'  => __( 'Optimization Detective', 'optimization-detective' ),
+			'meta'   => array(
+				'onclick' => 'document.body.classList.toggle("od-debug");',
+			),
+		)
+	);
+}
+
+add_action( 'admin_bar_menu', 'od_debug_add_admin_bar_menu_item', 100 );
+
+/**
+ * Adds inline JS & CSS for debugging.
+ */
+function od_debug_add_assets(): void {
+	if ( ! current_user_can( 'customize' ) && ! wp_is_development_mode( 'plugin' ) ) {
+		return;
+	}
+
+	if ( ! is_admin_bar_showing() ) {
+		return;
+	}
+
+	$tag_visitor_registry = new OD_Tag_Visitor_Registry();
+
+	/** This action is documented in optimization.php. */
+	do_action( 'od_register_tag_visitors', $tag_visitor_registry );
+
+	$group_collection = od_get_group_collection( $tag_visitor_registry );
```

Comment:
Note: This could be added as part of the tag visitor instead. It would already have access to the `$group_collection` in that context, and the visitor could add the HTML to the document via `$context->processor->append_body_html( $markup )`. Granted, it doesn't really make sense as a tag visitor either. So nevermind. There should be helpers to obtain the current group collection outside the context of a tag visitor.

Reply:
But there is a tag visitor in `Optimization_Detective_Debug_Tag_Visitor`. So it could be moved there.